cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/RackAwareDistributionGoal.java

```diff
@@ -0,0 +1,353 @@
+import com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException;
+import com.linkedin.kafka.cruisecontrol.model.Broker;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.Replica;
+import com.linkedin.kafka.cruisecontrol.model.ReplicaSortFunctionFactory;
+import com.linkedin.kafka.cruisecontrol.model.SortedReplicasHelper;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.function.Function;
+
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.GoalUtils.replicaSortName;
+import static java.util.Collections.max;
+import static java.util.Collections.min;
+
+
+/**
+ * HARD GOAL: Generate replica movement proposals to evenly distribute replicas over racks.
+ *
+ * This is a relaxed version of {@link RackAwareGoal}. Contrary to {@link RackAwareGoal}, as long as replicas of each
+ * partition can achieve a perfectly even distribution across the racks, this goal lets placement of multiple replicas
+ * of a partition into a single rack.
+ *
+ * <p>
+ *   For example, suppose a topic with 1 partition has 4 replicas in a cluster with 2 racks. Then the following
+ *   distribution will be acceptable by this goal (but would be unacceptable by {@link RackAwareGoal}):
+ * </p>
+ *
+ * <pre>
+ * Rack A                    | Rack B
+ * -----------------------------------------------------
+ * Broker1-rack A  replica-0 | Broker2-rack B
+ * Broker3-rack A            | Broker4-rack B replica-1
+ * Broker5-rack A  replica-2 | Broker6-rack B
+ * Broker7-rack A            | Broker8-rack B replica-3
+ * </pre>
+ *
+ * <p>
+ *   However, this goal will yield an {@link OptimizationFailureException} for the same partition in the following
+ *   cluster due to the lack of a second broker to place a replica of this partition in {@code Rack B}:
+ * </p>
+ *
+ * <pre>
+ * Rack A                    | Rack B
+ * -----------------------------------------------------
+ * Broker1-rack A  replica-0 | Broker2-rack B replica-1
+ * Broker3-rack A  replica-3 |
+ * Broker5-rack A  replica-2 |
+ * </pre>
+ */
+public class RackAwareDistributionGoal extends AbstractRackAwareGoal {
+  private BalanceLimit _balanceLimit;
+
+  /**
+   * Constructor for Rack Aware Distribution Goal.
+   */
+  public RackAwareDistributionGoal() {
+  }
+
+  /**
+   * Package private for unit test.
+   */
+  RackAwareDistributionGoal(BalancingConstraint constraint) {
+    _balancingConstraint = constraint;
+  }
+
+  @Override
+  protected boolean doesReplicaMoveViolateActionAcceptance(ClusterModel clusterModel,
+                                                           Function<ClusterModel, Replica> sourceReplicaFunction,
+                                                           Function<ClusterModel, Broker> destinationBrokerFunction) {
```

Comment:
Why are functions ClusterModel -> replica and ClusterModel -> Broker passed in here rather than the return value of the function?  There is only one parameter for each of these functions and that parameter is also passed into doesReplicaMoveViolateActionAcceptance.  So the method could be
(ClusterModel clusterModel, Replica sourceReplica, Broker destinationBroker)  No?

These functions are also only called once.

Reply:
Indeed there is no need for this extra complexity -- thanks!